core/src/main/kotlin/com/strumenta/kolasu/ids/DeclarativeNodeIdProvider.kt

```diff
@@ -1,67 +1,145 @@
+    /**
+     * Add an information [message] during the rule execution.
+     * @param message the information message content
+     **/
+    fun info(message: String)
+
+    /**
+     * Add a warning [message] during the rule execution
+     **/
+    fun warning(message: String)
+
+    /**
+     * Add an error [message] during the rule execution
+     **/
+    fun error(message: String)
+}
+
+private class ConfigurableNodeIdProvider : NodeIdProviderConfigurationApi, NodeIdProvider {
+
+    private val rules: MutableList<ConfigurableNodeIdProviderRule<*>> = mutableListOf()
+
+    override fun <NodeTy : Node> idFor(
+        nodeType: KClass<out NodeTy>,
+        rule: NodeIdProviderConfigurationRuleApi.(NodeIdProviderConfigurationRuleContext<out NodeTy>) -> String
+    ) {
+        this.rules.add(ConfigurableNodeIdProviderRule(nodeType, rule))
+    }
+
+    override fun id(kNode: Node): String {
+        return this.rules.sorted()
+            .firstOrNull { it.canBeInvokedWithReceiver(kNode::class) }
+            ?.invoke(this, kNode, mutableListOf())
+            ?: throw RuntimeException("Error while retrieving node id - no compatible rule found")
+    }
+}
+
+private class ConfigurableNodeIdProviderRule<NodeTy : Node>(
     private val nodeType: KClass<NodeTy>,
-    private val specification: NodeIdProvider.(NodeTy) -> String
-) : Comparable<DeclarativeNodeIdProviderRule<*>>, (NodeIdProvider, Node) -> String {
-    override fun invoke(
-        nodeIdProvider: NodeIdProvider,
-        node: Node
-    ): String {
-        @Suppress("UNCHECKED_CAST")
-        return nodeIdProvider.specification(node as NodeTy)
+    private val configuration: NodeIdProviderConfigurationRuleApi
+    .(NodeIdProviderConfigurationRuleContext<NodeTy>) -> String
+) : NodeIdProviderConfigurationRuleApi,
+    (NodeIdProvider, Node, MutableList<Issue>) -> String,
+    Comparable<ConfigurableNodeIdProviderRule<*>> {
+
+    private lateinit var context: NodeIdProviderConfigurationRuleContext<NodeTy>
+    private lateinit var issues: MutableList<Issue>
+
+    override fun info(message: String) {
+        this.issue(message, INFO)
+    }
+
+    override fun warning(message: String) {
+        this.issue(message, WARNING)
+    }
+
+    override fun error(message: String) {
+        this.issue(message, ERROR)
     }
 
-    fun canBeInvokedWith(type: KClass<*>): Boolean {
-        return this.nodeType.isSuperclassOf(type)
+    private fun issue(message: String, severity: IssueSeverity) {
+        this.issues.add(Issue.semantic(message, severity, this.context.node.position))
     }
 
-    override fun compareTo(other: DeclarativeNodeIdProviderRule<*>): Int {
+    override fun compareTo(other: ConfigurableNodeIdProviderRule<*>): Int {
         return when {
             this.nodeType.isSuperclassOf(other.nodeType) -> 1
-            this.nodeType.isSubclassOf(other.nodeType) -> -1
+            other.nodeType.isSuperclassOf(this.nodeType) -> -1
```

Comment:
Can we use also interfaces as nodeTypes? If so, would this work?

Reply:
Given the definition `NodeTy: Node`, I do not think it would work with interfaces. In order for this to work with interfaces, we would need to remove the type constraint.

Reply:
Right, let's forget about this then. This may change in Kolasu 1.6 where we have an interface to represent anything that can be a Node

Reply:
I agree, this is just a temporary solution that will radically change in Kolasu 1.6